testing.md

```diff
@@ -0,0 +1,91 @@
+# How to test changes
+This document assumes you will be using personal quay.io registry to push bundle and operator images to avoid accidentally changing production registries
+```
+export QUAY_REGISTRY=quay.io/<your_namespace>
+```
+
+Also create a CatalogSource
+```
+cat <<EOF | tee >(oc apply -f -) | cat
+apiVersion: operators.coreos.com/v1alpha1
+kind: CatalogSource
+metadata:
+  name: opencloud-operators
+  namespace: openshift-marketplace
+spec:
+  displayName: opencloud-operators
+  publisher: IBM
+  sourceType: grpc
+  image: quay.io/opencloudio/ibm-common-service-catalog:3.7.1
+  updateStrategy:
+    registryPoll:
+      interval: 45m
+EOF
+```
+
+If pushing an image to your quay.io registry with its first tag, the image repository may be set to private by default, so you will need to login to quay.io and change the repository settings to public in order for your cluster to pull the image.
+
+The actual testing consist of:
+1. verify common-service operator pod is Running
+2. verify namespace scope operator and ODLM operator are installed
+3. verify secretshare and common-service-webhook pods are running
+4. verify functionality of common-service operator such as:
+    - size profile
+
+## Test fresh installation with bundle
+1. add features/fixes to repo
+2. change `image` value in config/manager/manager.yaml to `quay.io/<your_namespace>/common-service-operator:dev`
+   - and any of the operand image values if necessary
+3. build operator with changes
+```
+make build-dev-image
+```
+4. build bundle containing changes and bundle image
+```
+make bundle-manifests RELEASE_VERSION=99.99.99
+make build-bundle-image RELEASE_VERSION=dev
+```
+5. deploy operator using bundle format
+```
+make run-bundle RELEASE_VERSION=dev
+```
+6. run tests
+7. clean up
+```
+make cleanup-bundle
+```
+
+## Test upgrade with bundle
+Similar to fresh installation test except you will first deploy the operator/bundle of the most recent release without any changes, i.e. the operator/bundle code from the most recent commit in master branch
+
+1. change `image` value in config/manager/manager.yaml to `quay.io/<your_namespace>/common-service-operator:3.7.1`
```

Comment:
It can't test the CSV change during the upgrade, right?

Reply:
`operator-sdk run bundle-upgrade` command will test CSV changes. The `run bundle` and `run bundle-upgrade` commands create a local CatalogSource and actually create a Subscription for the operator.

Reply:
Yes, but from my understanding, these steps only replace the images used in csv. Both 3.7.1 and dev image are using the same CSV template, which is the latest version. So we miss the change between the 3.7.1 version CSV and the dev version CSV template.

Since we are using the CSV to keep some configurations, this upgrade testing ignores these configuration updates.

Reply:
Steps 4 - 7 would be where the developer can make changes to the CSV. And the bundle is re-created with the new CSV https://github.com/IBM/ibm-common-service-operator/pull/431/files#diff-c7f6993de081c6827242e2bb49530eb5cb078a7900d9d9391fdeef2ac9060970R71-R82

Reply:
From my understanding, steps 1-3 should happen on branch release-3.6/release3.7 and steps 4-7 should happen on the master branch.

Reply:
Oh, I see what you mean now. I think step 1 and 2 aren't actually needed. Because the bundle image for an existing release is already on quay.io, which should be referencing the correct operator image. So developer should just need to push the official bundle image to their registry and run 
```
make run-bundle
``` 
or optionally 
```
make run-bundle QUAY_REGISTRY=dev-registry BUNDLE_IMAGE_NAME=ibm-common-service-operator-bundle
```
Does that look right?

Reply:
Thanks, It sounds good

Reply:
new commits should resolve this